File: iOS/Chatter/SocketHelper.m
//
//  SocketHelper.m
//  Chatter
//
//  Created by Steven on 4/6/13.
//  Copyright (c) 2013 Steven. All rights reserved.
//

#import "DataHelper.h"
#import "SocketHelper.h"
#import "ChannelHelper.h"

#import "SocketIOPacket.h"

@implementation SocketHelper

static SocketHelper *instance = nil;

+ (SocketHelper *)getInstance
{
    // Implement singleton pattern
    if (instance == nil)
    {
        instance = [[super allocWithZone:nil] init];
    }

    return instance;
}

+ (id)allocWithZone:(NSZone *)zone
{
    return [self getInstance];
}

@synthesize client;

- (void)connect
{
    client = [[SocketIO alloc] initWithDelegate:self];
    [client connectToHost:@"chatter-7482.onmodulus.net" onPort:80];
}

- (SocketIO *)getClient
{
    return client;
}

- (void)socketIODidConnect:(SocketIO *)socket
{
    // Create data to send
    NSMutableDictionary *data = [NSMutableDictionary dictionary];
    [data setObject:[DataHelper getIdentifier] forKey:@"identifier"];

    [client sendEvent:@"channel_list" withData:data];
}

- (void)socketIO:(SocketIO *)socket didReceiveJSON:(SocketIOPacket *)packet
{
    NSLog(@"didReceiveJSON: %@", packet.data);
}

- (void)socketIO:(SocketIO *)socket didReceiveEvent:(SocketIOPacket *)packet
{
    NSLog(@"didReceiveEvent() >>> data: %@", packet.data);

    // The global event 'nickchanged' is handled seperately.
    if ([packet.name isEqualToString:@"nickchanged"])
    {
        NSUserDefaults *defaults  = [NSUserDefaults standardUserDefaults];
        NSString *local_nickname  = [defaults objectForKey:@"nickname"];
        NSString *server_nickname = [[[packet.dataAsJSON objectForKey:@"args"] objectAtIndex:0] objectForKey:@"nickname"];

        if (!local_nickname || [local_nickname isEqualToString:@""])
        {
            // Save te given nickname
            NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
            [defaults setObject:local_nickname forKey:@"nickname"];
        }
        else if (![local_nickname isEqualToString:server_nickname])
        {
            // Notify the server of the local nickname
            NSMutableDictionary *data = [NSMutableDictionary dictionary];
            [data setObject:local_nickname forKey:@"nickname"];
            [client sendEvent:@"nickchange" withData:data];
        }

        return;
    }

    // The global event 'channel_list' is handled seperately.
    if ([packet.name isEqualToString:@"channel_list"])
    {
        // Retrieve JSON output
        NSDictionary *data = [[[packet.dataAsJSON objectForKey:@"args"] objectAtIndex:0] objectForKey:@"channels"];

        if (data)
        {
            // Loop through channels and add them
            for (NSDictionary *dict in data)
            {
                NSString *name = [dict objectForKey:@"channel"];
                [ChannelHelper addChannel:[[Channel alloc] initWithName:name]];
            }

            // Notify subscribers
            [ChannelHelper notifySubscribers];
        }

        return;
    }

    // Every other event is attached to a channel. Retrieve it.
    NSMutableDictionary *data = [[packet.dataAsJSON objectForKey:@"args"] objectAtIndex:0];
    Channel *channel = [ChannelHelper getChannel:[data objectForKey:@"channel"]];

    // There is no reason to go any further if there isn't a channel
    if (!channel) return;

    if ([packet.name isEqualToString:@"log"])
    {
        // Retrieve the history
        NSArray *history = [[[data objectForKey:@"history"] reverseObjectEnumerator] allObjects];

        // Loop through all events and add them (Backwards ofcourse)
        for (NSDictionary *event in history)
        {
            // Retrieve the event type
            NSString *type = [event objectForKey:@"type"];

            if ([type isEqualToString:@"msg"])
            {
                // Add event to channel
                NSString *nickname = [event objectForKey:@"nickname"];
                NSString *content  = [event objectForKey:@"message"];
                [channel addEvent:[NSString stringWithFormat:@"[%@] %@", nickname, content]];
            }
            else if ([type isEqualToString:@"join"])
            {
                // Add event to channel
                NSString *nickname = [event objectForKey:@"nickname"];
                [channel addEvent:[NSString stringWithFormat:@"%@ joined the room.", nickname]];
            }
            else if ([type isEqualToString:@"part"])
            {
                // Add event to channel
                NSString *nickname = [event objectForKey:@"nickname"];
                [channel addEvent:[NSString stringWithFormat:@"%@ left the room.", nickname]];
            }
            else if ([type isEqualToString:@"nickchange"])
            {
                // Add event to channel
                NSString *old_nick = [event objectForKey:@"message"];
                NSString *new_nick = [event objectForKey:@"nickname"];
                [channel addEvent:[NSString stringWithFormat:@"%@ renamed to %@", old_nick, new_nick]];
            }
        }

        // Update subscribers
        [ChannelHelper notifySubscribers:channel];
    }
    else if ([packet.name isEqualToString:@"msg"])
    {
        // Add event to channel
        NSString *nickname = [data objectForKey:@"nickname"];
        NSString *content  = [data objectForKey:@"content"];
        [channel addEvent:[NSString stringWithFormat:@"[%@] %@", nickname, content]];

        // Update subscribers
        [ChannelHelper notifySubscribers:channel];
    }
    else if ([packet.name isEqualToString:@"user_list"])
    {
        // TODO
    }
    else if ([packet.name isEqualToString:@"user_joined"])
    {
        // Add event to channel
        NSString *nickname = [data objectForKey:@"nickname"];
        [channel addEvent:[NSString stringWithFormat:@"%@ joined the room.", nickname]];

        // Update subscribers
        [ChannelHelper notifySubscribers:channel];
    }
    else if ([packet.name isEqualToString:@"user_parted"])
    {
        // Add event to channel
        NSString *nickname = [data objectForKey:@"nickname"];
        [channel addEvent:[NSString stringWithFormat:@"%@ left the room.", nickname]];

        // Update subscribers
        [ChannelHelper notifySubscribers:channel];
    }
    else if ([packet.name isEqualToString:@"user_nickchanged"])
    {
        // Add event to channel
        NSString *old_nick = [data objectForKey:@"old_nick"];
        NSString *new_nick = [data objectForKey:@"new_nick"];
        [channel addEvent:[NSString stringWithFormat:@"%@ renamed to %@", old_nick, new_nick]];

        // Update subscribers
        [ChannelHelper notifySubscribers:channel];
    }
}

- (void)socketIO:(SocketIO *)socket didReceiveMessage:(SocketIOPacket *)packet
{
    NSLog(@"didReceiveMessage: %@", packet.data);
}

@end

File: Android/src/nl/stevenhack/chatter/helpers/SocketHelper.java
package nl.stevenhack.chatter.helpers;

import java.net.URI;

import android.content.SharedPreferences;
import android.os.Looper;
import android.preference.PreferenceManager;
import android.util.Log;

import nl.stevenhack.chatter.models.Channel;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.codebutler.android_websockets.SocketIOClient;

/**
 * 
 * @author Steven
 *
 * This Helper class handles the Socket.IO connection with the server.
 * All incoming and outgoing events go through this class.
 * Incoming events are redirected towards the correct target.
 * 
 */
public class SocketHelper implements SocketIOClient.Handler {

	/**
	 * This variable is used for sending notifications to the UI thread.
	 * It has to be final because of memory references.
	 */
	//private final Channel channel = null;

	/**
	 * Singleton instance of the SocketIOClient class.
	 */
	private static SocketIOClient client = null;

	/**
	 * Returns the singleton instance of the SocketIOClient class.
	 * If it is not created yet, it does so before returning.
	 */
	public static SocketIOClient getInstance() {
		if (client == null)
			client = new SocketIOClient(URI.create("http://chatter-7482.onmodulus.net"), new SocketHelper());

		return client;
	}

	@Override
	public void onConnect() {
		// Retrieve a list of all subscribed channels
		try {
			client.emit("channel_list", 
				new JSONArray().put(
					new JSONObject().put("identifier", DataHelper.getIdentifier())
				)
			);
		}
		catch (JSONException e) {
			e.printStackTrace();
			Log.d("SocketHelper", String.format("Could not retrieve channels: %s", e.getMessage()));
		}
	}

	@Override
	public void on(String event, JSONArray arguments) {
		// The global event 'nickchanged' is handled seperately.
		if (event.equals("nickchanged")) {
			SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(DataHelper.getContext());
			String local_nickname = prefs.getString("preference_nickname", "");
			String server_nickname = DataHelper.getStringFromJSONArray(arguments, "nickname");

			if (local_nickname.equals("")) {
				SharedPreferences.Editor editor = prefs.edit();
				editor.putString("preference_nickname", server_nickname);
				editor.commit();
			}
			else if (!local_nickname.equals(server_nickname)) {
				// Notify the server of the local nickname
				try {
					client.emit("nickchange", 
						new JSONArray().put(
							new JSONObject().put("nickname", local_nickname)
						)
					);
				}
				catch (JSONException e) {
					e.printStackTrace();
					Log.d("SocketHelper", String.format("Could not update nickname: %s", e.getMessage()));
				}
			}

			return;
		}

		// The global event 'channel_list' is handled seperately.
		if (event.equals("channel_list")) {
			try {
				// Retrieve channels
				JSONObject data = DataHelper.getObjectFromJSONArray(arguments);
				JSONArray channels = (JSONArray) data.get("channels");

				// Loop through channels and add them
				for (int i = 0; i < channels.length(); i++) {
					String name = DataHelper.getStringFromJSONObject(channels.getJSONObject(i), "channel");
					ChannelHelper.addChannel(new Channel(name));
				}

				// Update subscribers on main thread
				android.os.Handler handler = new android.os.Handler(Looper.getMainLooper());
				handler.post(new Runnable() {
					public void run() {
						ChannelHelper.notifySubscribers();
					}
				});
			}
			catch (JSONException e) {
				e.printStackTrace();
			}

			return;
		}

		// Every other event is attached to a channel. Retrieve it.
		JSONObject data = DataHelper.getObjectFromJSONArray(arguments);
		String name = DataHelper.getStringFromJSONObject(data, "channel");

		// Retrieve the channel from list of active channels
		// Make this variable final so it can be referenced to the UI Thread
		final Channel channel = ChannelHelper.getChannel(name);

		// If the channel could not be retrieved, there is no reason to go any further
		if (channel == null) {
			Log.d("SocketHelper", String.format("No channel given in event '%s': %s", event, arguments.toString()));
			return;
		}

		// Direct the event to the channel in a proper way
		if (event.equals("log")) {
			// Retrieve the history array
			JSONArray history = DataHelper.getArrayFromJSONObject(data, "history");

			// Loop through all events and add them (Backwards ofcourse)
			for (int i = history.length() - 1; i >= 0; i--) {
				JSONObject log = DataHelper.getObjectFromJSONArray(history, i);

				String type      = DataHelper.getStringFromJSONObject(log, "type");
				//String createdAt = DataHelper.getStringFromJSONObject(log, "createdAt");

				if (type.equals("msg")) {
					String message  = DataHelper.getStringFromJSONObject(log, "message");
					String nickname = DataHelper.getStringFromJSONObject(log, "nickname");
					channel.addEvent("[" + nickname + "] " + message);
				}
				else if (type.equals("join")) {
					String nickname = DataHelper.getStringFromJSONObject(log, "nickname");
					channel.addEvent(nickname + " joined the room.");
				}
				else if (type.equals("part")) {
					String nickname = DataHelper.getStringFromJSONObject(log, "nickname");
					channel.addEvent(nickname + " left the room.");
				}
				else if (type.equals("nickchange")) {
					String old_nick = DataHelper.getStringFromJSONObject(data, "message");
					String new_nick = DataHelper.getStringFromJSONObject(data, "nickname");
					channel.addEvent(old_nick + " renamed to " + new_nick);
				}
			}

			// Update subscribers on main thread
			android.os.Handler handler = new android.os.Handler(Looper.getMainLooper());
			handler.post(new Runnable() {
				public void run() {
					ChannelHelper.notifySubscribers(channel);
				}
			});
		}
		else if (event.equals("msg")) {
			String nickname = DataHelper.getStringFromJSONObject(data, "nickname");
			String content  = DataHelper.getStringFromJSONObject(data, "content");
			channel.addEvent("[" + nickname + "] " + content);

			// Update subscribers on main thread
			android.os.Handler handler = new android.os.Handler(Looper.getMainLooper());
			handler.post(new Runnable() {
				public void run() {
					ChannelHelper.notifySubscribers(channel);
				}
			});
		}
		else if (event.equals("user_list")) {
			// TODO
		}
		else if (event.equals("user_joined")) {
			String nickname = DataHelper.getStringFromJSONObject(data, "nickname");
			channel.addEvent(nickname + " joined the room.");

			// Update subscribers on main thread
			android.os.Handler handler = new android.os.Handler(Looper.getMainLooper());
			handler.post(new Runnable() {
				public void run() {
					ChannelHelper.notifySubscribers(channel);
				}
			});
		}
		else if (event.equals("user_parted")) {
			String nickname = DataHelper.getStringFromJSONObject(data, "nickname");
			channel.addEvent(nickname + " left the room.");

			// Update subscribers on main thread
			android.os.Handler handler = new android.os.Handler(Looper.getMainLooper());
			handler.post(new Runnable() {
				public void run() {
					ChannelHelper.notifySubscribers(channel);
				}
			});
		}
		else if (event.equals("user_nickchanged")) {
			String old_nick = DataHelper.getStringFromJSONObject(data, "old_nick");
			String new_nick = DataHelper.getStringFromJSONObject(data, "new_nick");
			channel.addEvent(old_nick + " renamed to " + new_nick);

			// Update subscribers on main thread
			android.os.Handler handler = new android.os.Handler(Looper.getMainLooper());
			handler.post(new Runnable() {
				public void run() {
					ChannelHelper.notifySubscribers(channel);
				}
			});
		}
	}

	@Override
	public void onJSON(JSONObject json) { Log.d("SocketHelper", String.format("New JSON: %s", json.toString())); }

	@Override
	public void onMessage(String message) { Log.d("SocketHelper", String.format("New message: %s", message)); }

	@Override
	public void onDisconnect(int code, String reason) {
		Log.d("SocketHelper", String.format("Disconnected! Code: %d Reason: %s", code, reason));
	}

	@Override
	public void onError(Exception error) {
		Log.e("SocketHelper", "Uknown error:", error);
	}
}
